optuna/samplers/grid.py

```diff
@@ -0,0 +1,177 @@
+
+            def objective(trial):
+                x = trial.suggest_uniform('x', -100, 100)
+                y = trial.suggest_int('y', -100, 100)
+                return x ** 2 + y ** 2
+
+            search_space = {
+                'x': [-50, 0, 50],
+                'y': [-99, 0, 99]
+            }
+            study = optuna.create_study(sampler=optuna.samplers.GridSampler(search_space))
+            study.optimize(objective, n_trials=3*3)
+
+    Note:
+
+        :class:`~optuna.samplers.GridSampler` does not take care of a parameter's quantization
+        specified by discrete suggest methods but just samples one of values specified in the
+        search space. E.g., in the following code snippet, either of ``-0.5`` or ``0.5`` is
+        sampled as ``x`` instead of an integer point.
+
+        .. testcode::
+
+            import optuna
+
+            def objective(trial):
+                # The following suggest method specifies integer points between -5 and 5.
+                x = trial.suggest_discrete_uniform('x', -5, 5, 1)
+                return x ** 2
+
+            # Non-int points are specified in the grid.
+            search_space = {'x': [-0.5, 0.5]}
+            study = optuna.create_study(sampler=optuna.samplers.GridSampler(search_space))
+            study.optimize(objective, n_trials=2)
+
+    Args:
+        search_space:
+            A dictionary whose key and value are a parameter name and the corresponding candidates
+            of values, respectively.
+    """
+
+    def __init__(self, search_space):
+        # type: (Dict[str, List[GridValueType]]) -> None
+
+        for param_name, param_values in search_space.items():
+            for value in param_values:
+                self._check_value(param_name, value)
+            search_space[param_name] = param_values[:]
+
+        self._search_space = collections.OrderedDict(
+            sorted(search_space.items(), key=lambda x: x[0]))
+        self._all_grids = list(itertools.product(*self._search_space.values()))
+        self._param_names = sorted(search_space.keys())
+        self._n_min_trials = len(self._all_grids)
+
+    def infer_relative_search_space(self, study, trial):
+        # type: (Study, FrozenTrial) -> Dict[str, BaseDistribution]
+
+        return {}
+
+    def sample_relative(self, study, trial, search_space):
+        # type: (Study, FrozenTrial, Dict[str, BaseDistribution]) -> Dict[str, Any]
+
+        # Instead of returning param values, GridSampler puts the target grid id as a system attr,
+        # and the values are returned from `sample_independent`. This is because the distribution
+        # object is hard to get at the beginning of trial, while we need the access to the object
+        # to validate the sampled value.
+
+        unvisited_grids = self._get_unvisited_grid_ids(study)
+
+        if len(unvisited_grids) == 0:
+            raise ValueError('All grids have been evaluated. If you want to avoid this error, '
```

Comment:
I wanna discuss whether GridSampler should fallback to RandomSampler or not before approving this PR.

Reply:
Now I feel it should not fallback to stop `study.optimize()` immediately. It's a bother to consider the appropriate `n_trials` from grids.

Reply:
Let me make sure of your points:
- the sampler should not switch to random sampling;
- the sampler should be more user-friendly so that it automatically stop the study after finishing all grids.

Are they correct?

Reply:
Given they are correct, I agree both of them.

For the latter point, we'd need a new feature to suspend a study (e.g., something like `optuna.exceptions.StopStudy`). I think the task is a bit large to be included in this PR and can be addressed later in another PR.

Reply:
> Are they correct?

Yes.

> For the latter point, we'd need a new feature to suspend a study (e.g., something like optuna.exceptions.StopStudy). I think the task is a bit large to be included in this PR and can be addressed later in another PR.

Totally agree. I feel it's better to discuss in another PR.